pkg/cache/sql/informer/factory/informer_factory.go

```diff
@@ -25,15 +27,16 @@ const EncryptAllEnvVar = "CATTLE_ENCRYPT_CACHE_ALL"
 
 // CacheFactory builds Informer instances and keeps a cache of instances it created
 type CacheFactory struct {
-	informerCreateLock sync.RWMutex
-	wg                 wait.Group
-	dbClient           DBClient
-	stopCh             chan struct{}
-	encryptAll         bool
+	wg          wait.Group
+	dbClient    DBClient
+	stopCh      chan struct{}
+	stopChMutex sync.RWMutex
+	encryptAll  bool
 
 	newInformer newInformer
 
-	cache map[schema.GroupVersionKind]*informer.Informer
+	informers       sync.Map
+	informerMutexes sync.Map
```

Comment:
Curious as to why the use of `sync.Map` instead of a simple map+mutex. There's a bit of added complexity with map with the different methods than a regular map, the fact that it's basically a `map[any]any` so needs casting everywhere and it's not clear when looking at it what the type of key or value is.

Also feels like it would be easier to maintain using a single map instead of two that needs to be kept in sync. Something like `map[gvk]*guardedInformer` where the `guardedInformer` is just a mtuex+the informer. (If you can come up with a better name, even better :D )

WDYT?

Reply:
1. Yeah I can see how having to deal with two maps and two locks would end up not very readable. I wonder if a similar approach would have worked with a single lock and two maps. Basically, lock the maps with single lock, initialize both the RWMutex and the Informer if it doesn't exists yet, unlock the maps with single lock, lock the per-type lock, start the informer or something.. Anyway, hard to reason with all these locks.


2. Yeah, if you're opened to `map[gvk]*guardedInformer + Mutex` and it works for our use case then we should go with that, I think it's going to be more readable.

Reply:
I tried but failed to come up with a satisfying solution with one only map of `guardedInformer`s.

The reason for that is management of the time interval when:

- an informer for a given GVK has not been created yet (so a `guardedInformer.informer` is still nil) but
- the indexer-specific mutex has to be locked, to prevent multiple Informers may start at the same time for that GVK (so `guardedInformer.mutex` has to be non-nil)

Code-wise that corresponds to these lines: https://github.com/moio/lasso/blob/075de697fc1615403d6ca6d3f982a743673d3fca/pkg/cache/sql/informer/factory/informer_factory.go#L102-L117

Having both the informer and the informer mutex in the same map means I have to lock/unlock the map twice: first get the mutex (and set it if it does not exist - bringing in logic similar to [LoadOrStore](https://cs.opensource.google/go/go/+/refs/tags/go1.23.1:src/sync/map.go;l=227), which is not trivial), then to set the informer once it's created.

Reply:
To make it a bit more concrete, this is what I came up with:

https://github.com/moio/lasso/pull/1/files

To me that is subjectively harder to follow despite types, mainly because I am breaking the "Lock then defer" pattern.

The other problem it has is that [golang's RWMutex does not support lock upgrading](https://github.com/golang/go/issues/38859#issuecomment-623742366), which, IIUC, means that code has a risk of deadlocking if two goroutines enter [this line](https://github.com/moio/lasso/pull/1/files#diff-30ea1a028c2e55261cf0c7fd16267f209c53187a2fa4372afc2580001b9a668bR114).

Happy to hear your thoughts.

Reply:
How about something like this? [pkg/cache/sql/informer/factory/informer_factory.go#L95-L129](https://github.com/rancher/lasso/blob/913ec0e0ca62faed914029b9cb5178f7ccdcb06f/pkg/cache/sql/informer/factory/informer_factory.go#L95-L129)

My assumption (and let me know if I'm wrong) is that there's two things that takes time:
1. newInformer: should be pretty quick iirc as it's only creating a DB and setting up the tables
2. waiting for caches to sync: this one I expect to take the longest when there's a high number of objects

So with these assumptions, the code locks the global mutex for a very short amount of time, letting other GVK start at the same time. Then, each GVK is locked until informer is set and started (should be quick).

Finally, no locks are taken while we wait for the caches to sync.

Let me know your thoughts.

Reply:
I like your proposal - I was overcomplicating my life because `RWMutex`es, while a plain `Mutex` is probably as effective, and much more readable.

One problem with the code was `gi.informer` was never set - fixed now.

Another thing that's entirely opinable is the placement of `WaitForCacheSync`. That method spinlocks until the informer is ready, so in principle it can be outside of the lock as you put it. Nevertheless there is no real advantage in having it called by non-creating goroutines, so I moved it inside the lock. That had the nice side effect of allowing to use defer on the `gi.mutex`, which IMO adds a bit of readability.

Concluding this discussion and waiting for your next round of comments on the patch overall.

Reply:
>  Nevertheless there is no real advantage in having it called by non-creating goroutines, so I moved it inside the lock.

That's perfect :+1: it's better this way